proposals/BVAL-508.md

```diff
@@ -0,0 +1,257 @@
+
+## Problem
+
+It is useful to be able to apply constrains on elements contained in a so called _container_.
+Examples of containers are:
+
+* `Collection`
+* `Optional`
+* JavaFX's `Property`
+
+But today the constraints are applied on the container value itself and not its content.
+
+## Proposition
+
+### Containers
+
+Bean Validation will offer the notion of container of values to validate.
+Containers can contain one value, or several values.
+Extractors for specific containers can be implemented.
+
+TODO: should we differentiate single and plural?
+
+    // Draft proposal for the contract
+    public interface SingleContainerValueExtractor<CONTAINER,CONTAINED> {
+        CONTAINED extractValue(CONTAINER container);
+    }
+    
+    public interface ManyContainerValuesExtractor<CONTAINER,CONTAINED> {
+        Iterable<CONTAINED> extractValue(CONTAINER container);
+    }
+    
+    // Examples
+    
+    //TODO can T remain unspecified? or should it be ? or Object?
+    public class OptionalContainerExtractor<T> implements SingleContainerValueExtractor<Optional<T>,T> {
+        T extractValue(Optional<T> optional) { ... };
+    }
+
+Bean Validation will have out of the box support for containers for `Iterable` and `Map`.
+
+TODO: any other Java SE type to consider?
+TODO: should we support JavaFX Property<T>?
+
+Custom container extractors can be provided via:
+
+* property
+* factory
+* service locator
+
+TODO: refine how the container implementations are discovered.
+See [Hibernate Validator's feature](http://docs.jboss.org/hibernate/validator/5.2/reference/en-US/html_single/#section-value-handling) 
+
+#### When are containers used?
+
+When a property is of a known container, constraints declarations are explored.
+A typical example is a property using constraints on type parameters of `TYPE_USE` (`Collection<@Size String>`).
+
+If constraints are discovered, the unwrapping mechanism is used
+to read contained value(s) and potentially apply constraints on them (see below).
+
+By constraints, we also mean property traversal `@Valid`.
+
+Containers of containers are applied recursively if necessary.
+TODO: infinite loops?
+
+#### Which container unwrapper to use?
+
+The container used is the one corresponding to the most specific super type
+as defined in the BV resolution of constraint validators.
+
+### Constraints applied and containers
+
+Constraints declared on the type parameter of a type use will be applied to the contained value
+as extracted by the container logic.
+
+    // each String of the collection is validated for the regexp
+    Set<@Pattern(...) String> emails;
+
+By default, constraints declared on the container will apply to the container.
+This ensures backward compatibility.
+
+    // @Size is applied on the collection
+    @Size(min=5) List<Integer> ages;
+
+Extractors can specify that constraints declared on the container apply to the contained value(s);
+
+    @ConstraintsApplyTo(CONTAINED_VALUES)
```

Comment:
Ahh, I think I am with you know. It really stems from the fact as well that you try to unify collections (which were sort of a special case already) and other data containers. 

I somehow thought that it would make things "clearer" for the end user which is not the cases. For the end user there is still a bit of pixie dust applied, since in some cases the constraint will apply to the container (collections) and in some cases to the wrapped value (FX property). One would need to look into the implementation of the "extractor" to get the full picture. 

Either way, I think I am on board now. I guess concrete names can be discussed later. Other than that I think the proposal as a whole could need some more "hand holding" to make it easier to understand.